pkg/controllers/management/certsexpiration/certsexpiration.go

```diff
@@ -0,0 +1,176 @@
+	rkecluster "github.com/rancher/rke/cluster"
+	"github.com/rancher/rke/pki"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/cert"
+)
+
+type Controller struct {
+	ClusterLister   v3.ClusterLister
+	ClusterClient   v3.ClusterInterface
+	ClusterManager  *clustermanager.Manager
+	NamespaceClient v1.NamespaceInterface
+	CoreV1          v1.Interface
+}
+
+func Register(ctx context.Context, management *config.ManagementContext, manager *clustermanager.Manager) {
+	c := &Controller{
+		ClusterLister:   management.Management.Clusters("").Controller().Lister(),
+		ClusterClient:   management.Management.Clusters(""),
+		NamespaceClient: management.Core.Namespaces(""),
+		CoreV1:          management.Core,
+		ClusterManager:  manager,
+	}
+	management.Management.Clusters("").AddHandler(ctx, "certificate-expiration", c.sync)
+}
+
+func (c Controller) sync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if key == "" || cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if cluster.Spec.RancherKubernetesEngineConfig == nil {
+		return nil, nil
+	}
+	certsExpInfo := map[string]v3.CertExpiration{}
+
+	cluster, err := c.ClusterLister.Get("", key)
+	if err != nil {
+		return nil, err
+	}
+	certBundle, err := c.getClusterCertificateBundle(cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+	for certName, certObj := range certBundle {
+		info, err := getCertExpiration(certObj.CertificatePEM)
+		if err != nil {
+			continue
+		}
+		certsExpInfo[certName] = info
+	}
+	cluster.Status.CertificatesExpiration = certsExpInfo
+	return c.ClusterClient.Update(cluster)
+
+}
+
+func (c Controller) getClusterCertificateBundle(clusterName string) (map[string]pki.CertificatePKI, error) {
+	// cluster has a state file ?
+	store := clusterprovisioner.NewPersistentStore(c.NamespaceClient, c.CoreV1)
+	currentState, err := getRKECurrentStateFromStore(store, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	if currentState != nil {
+		cleanCertificateBundle(currentState.CertificatesBundle)
+		return currentState.CertificatesBundle, nil
+	}
+
+	// No state file, let's try get the certs from the user cluster
+	clusterContext, err := c.ClusterManager.UserContext(clusterName)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := getCertsFromUserCluster(clusterContext)
+	if err != nil {
+		return nil, err
+	}
+	cleanCertificateBundle(certs)
+	return certs, nil
+}
+
+func getRKECurrentStateFromStore(store cluster.PersistentStore, clusterName string) (*rkecluster.State, error) {
+	cluster, err := store.Get(clusterName)
+	if err != nil {
+		return nil, err
+	}
+	var fullState rkecluster.FullState
+	stateStr, ok := cluster.Metadata["fullState"]
```

Comment:
just a question, when will `fullState` not be present? in that case, do we need to look for it elsewhere?

Reply:
No, this is only available in 2.2.x clusters. If it's not found, it's 2.0.x or 2.1.x cluster, in that case we get the certificate bundle from the user cluster directly.